src/infra/core/src/ingest_service_impl.rs

```diff
@@ -7,44 +7,685 @@
             run_info_dir,
             cache_dir,
+            time_source,
+        }
+    }
+
+    async fn do_ingest(
+        &self,
+        dataset_ref: &DatasetRef,
+        options: IngestOptions,
+        fetch_override: Option<FetchStep>,
+        get_listener: impl FnOnce(&DatasetHandle) -> Option<Arc<dyn IngestListener>>,
+    ) -> Result<IngestResult, IngestError> {
+        let dataset_handle = self.dataset_repo.resolve_dataset_ref(&dataset_ref).await?;
+
+        self.dataset_action_authorizer
+            .check_action_allowed(&dataset_handle, auth::DatasetAction::Write)
+            .await?;
+
+        let dataset = self
+            .dataset_repo
+            .get_dataset(&dataset_handle.as_local_ref())
+            .await?;
+
+        let listener =
+            get_listener(&dataset_handle).unwrap_or_else(|| Arc::new(NullIngestListener));
+
+        // TODO: Remove this scan after we eliminate Spark ingest
+        let Some((_, polling_source_block)) = dataset
+            .as_metadata_chain()
+            .last_of_type::<SetPollingSource>()
+            .await
+            .int_err()?
+        else {
+            tracing::warn!(
+                %dataset_handle,
+                "Dataset does not define a polling source - considering up-to-date",
+            );
+
+            let result = IngestResult::UpToDate {
+                no_polling_source: true,
+                uncacheable: false,
+            };
+
+            listener.begin();
+            listener.success(&result);
+            return Ok(result);
+        };
+
+        // TODO: Temporarily we continue to default to Spark ingest, but use new engine
+        // if preprocessing query specifies datafusion
+        let engine = polling_source_block
+            .event
+            .preprocess
+            .as_ref()
+            .map(|Transform::Sql(t)| t.engine.as_str());
+
+        if engine == Some("datafusion") {
+            self.ingest_loop(IngestLoopArgs {
+                dataset_handle,
+                dataset,
+                options,
+                polling_source: polling_source_block.event,
+                fetch_override,
+                listener,
+            })
+            .await
+        } else {
+            self.legacy_spark_ingest(dataset_handle, dataset, options, fetch_override, listener)
+                .await
+        }
+    }
+
+    #[tracing::instrument(
+        level = "info",
+        skip_all,
+        fields(
+            dataset_handle = %args.dataset_handle,
+        )
+    )]
+    async fn ingest_loop(&self, args: IngestLoopArgs) -> Result<IngestResult, IngestError> {
+        let ctx = self.new_session_context();
+
+        let mut data_writer = DataWriterDataFusion::builder(args.dataset.clone(), ctx.clone())
+            .with_metadata_state_scanned()
+            .await?
+            .build()
+            .await?;
+
+        let mut iteration = 0;
+        let mut combined_result = None;
+        loop {
```

Comment:
As I understand, an iteration per query?

Reply:
Some data sources (currently only `filesGlob`) will return `has_more: true` to say "call me again - I have more data", so the ingest will loop, committing each chunk, until it exhausts the source.